src/System.Net.Primitives/tests/FunctionalTests/IPAddressTest.cs

```diff
@@ -242,23 +247,55 @@ public static void Equals_Compare_Success()
             Assert.False(ip4.GetHashCode().Equals(ip6.GetHashCode()));
         }
 
+        [Theory]
+        [MemberData(nameof(GetValidIPAddresses))]
+        public static void GetHashCode_ValidIPAddresses_Success(IPAddress ipAddress)
+        {
+            ipAddress.GetHashCode();
+        }
+
+        private static IEnumerable<object[]> GetValidIPAddresses()
+        {
+            return IPAddressParsing.ValidIpv4Addresses
+                .Concat(IPAddressParsing.ValidIpv6Addresses)
+                .Select(array => new object[] {IPAddress.Parse((string)array[0])});
+        }
+
+        [Theory]
+        [MemberData(nameof(GeneratedIPAddresses))]
+        public static void GetHashCode_VariousIPAddressGenerators_Success(IPAddress ipAddress)
+        {
+            ipAddress.GetHashCode();
+        }
+
+        public static readonly object[][] GeneratedIPAddresses =
+        {
+            new object[] {IPAddress.Parse(ipV4AddressString1)},
+            new object[] {IPAddress.Parse(ipV6AddressString)},
+            new object[] {new IPAddress(MinAddress)},
+            new object[] {new IPAddress(MaxAddress)},
+            new object[] {new IPAddress(ipV4AddressBytes)},
+            new object[] {new IPAddress(ipV6AddressBytes1)},
+            new object[] {new IPAddress(ipV6AddressBytes1, MinScopeId)},
+        };
+
 #pragma warning disable 618
-         [Fact]
-         public static void Address_Property_Failure()
-         {
-             IPAddress ip1 = IPAddress.Parse("fe80::200:f8ff:fe21:67cf");
-             Assert.Throws<SocketException>(() => ip1.Address);
-         }
- 
-         [Fact]
-         public static void Address_Property_Success()
-         {
-             IPAddress ip1 = IPAddress.Parse("192.168.0.9");
-             //192.168.0.10
-             long newIp4Address = 192 << 24 | 168 << 16 | 0 << 8 | 10;
-             ip1.Address = newIp4Address;
-             Assert.Equal("10.0.168.192" , ip1.ToString());
-         }
+        [Fact]
+        public static void Address_Property_Failure()
+        {
+            IPAddress ip1 = IPAddress.Parse("fe80::200:f8ff:fe21:67cf");
+            Assert.Throws<SocketException>(() => ip1.Address);
+        }
+
+        [Fact]
+        public static void Address_Property_Success()
+        {
+            IPAddress ip1 = IPAddress.Parse("192.168.0.9");
+            //192.168.0.10
+            long newIp4Address = 192 << 24 | 168 << 16 | 0 << 8 | 10;
+            ip1.Address = newIp4Address;
+            Assert.Equal("10.0.168.192", ip1.ToString());
```

Comment:
is there a test anywhere that setting .Address like this also resets the hash code?

of course you can't check the hash code changes, but you can check that the new hash code is the same as a fresh IPAddress created over the new address

Reply:
I think this bug already existed for ipv6, and now it is for ipv4 as well. Setting Address should clear _hashCode.

Reply:
I don't think this bug existed for ipv6 because you can't set `Address` for ipv6. However, i do think there's a similar bug with `ScopeId`. You can set it for ipv6 and it's part of the string representation (hence, also part of the current hash code) but the cached `_toString` doesn't get reset when `ScopeId` is set.
I now reset both `_toString` and `_hashCode` in the private address and scope id setters. I hope that's ok.